### DISCUSSION TITLE
Streaming of native chain tokens

### DISCUSSION THREAD

**Discussion by razgraf** [sablier-labs/evm-monorepo#41] [Ideas]
Our protocol allows for streaming of ERC20 tokens. While this reduces complexity to only having to handle one type of asset, we lack the possibility to stream chain native assets, such as ETH (on Ethereum), MATIC (on Polygon) or most recently requested RON (on Ronin).

With v1 because we don't have `onBehalf/from` variants for our methods it would be hard to achieve wrap+create, unwrap+withdraw. With v2 tho, this will be possible with a custom router that has payable functions and chains these transactions.

**Comment by PaulRBerg** [comment]
I agree that this would be a nice thing to have, but I would classify it as a priority 3 or a priority 4.

> With v2 tho, this will be possible with a custom router that has payable functions and chains these transactions.

Yes, possible to create streams. But at withdrawal time, it would still be the wrapped token version that the recipient receives.

**Comment by PaulRBerg** [comment]
Possible implementations for this, ranked from most to least plausible:

1. Wrap the tokens in a vault method and add an if statement in the `cancelInternal` and `withdrawInternal` functions that checks whether the ERC-20 token is the wrapped version of the underlying chain (e.g. WETH on Ethereum). If it is, unwrap the wrapped tokens and send them to the sender/ recipient.
2. Add a counterpart native asset function for every Sablier V2 function. Something like `createNative`, `withdrawNative`, etc. This would of course involve a lot of work.
3. Add a counterpart native asset streaming contract for every Sablier V2 contract. Something like `SablierV2NativeLinear`, `SablierV2CliffLinear`, etc. This would of course involve a lot of work, and it would be overkill.

**Comment by PaulRBerg** [comment]
Unfortunately, we really don't have time for this now.

It would be overkill, since users can just stream `WETH`, `WMATIC`, etc. Optionally, we might me able to add a simple converter function in the periphery (as part of a proxy target contract), but we definitely should not provide any counterparty native asset function or contracts.

Locking.

**Comment by PaulRBerg** [comment]
This is now possible thanks to our recent decision to employ the [proxy approach](https://github.com/sablierhq/v2-periphery/discussions/11).

**Comment by PaulRBerg** [comment]
For future reference - the `SablierV2ProxyTarget` in V2 Periphery has several functions that enable a wrap-and-create functionality:

- [`wrapAndCreateWithDurations`](https://github.com/sablier-labs/v2-periphery/blob/f859815e798d101568fbfbfc0165c72aea811b16/src/SablierV2ProxyTarget.sol#L352C14-L352C40)
- [`wrapAndCreateWithRange`](https://github.com/sablier-labs/v2-periphery/blob/f859815e798d101568fbfbfc0165c72aea811b16/src/SablierV2ProxyTarget.sol#L376C14-L376C36)
- [`wrapAndCreateWithDeltas`](https://github.com/sablier-labs/v2-periphery/blob/f859815e798d101568fbfbfc0165c72aea811b16/src/SablierV2ProxyTarget.sol#L579C14-L579C37)
- [`wrapAndCreateWithMilestones`](https://github.com/sablier-labs/v2-periphery/blob/f859815e798d101568fbfbfc0165c72aea811b16/src/SablierV2ProxyTarget.sol#L603C14-L603C41)
